### ISSUE TITLE
Align Git short aliases with Oh My Zsh

### ISSUE THREAD

**Issue by bittner** [Bash-it/bash-it#1191]
I've noticed that the short Git aliases in Bash-it do not match [the ones defined by the Oh My Zsh](https://github.com/robbyrussell/oh-my-zsh/blob/master/plugins/git/git.plugin.zsh#L36-L39) project.

Wouldn't it be a good idea to match them? This would make it easier for users to switch between Bash and Zsh.

**Comment by bittner** [comment]
Any hard feelings about this? Should I try to prepare a PR for [git.aliases.bash](https://github.com/Bash-it/bash-it/blob/master/aliases/available/git.aliases.bash)?

**Comment by nwinkler** [comment]
Before you embark on this, could you outline the main differences? To be honest, I think backwards-compatibility (and people's muscle memory) is more important than using the same aliases as Oh My Zsh.

If you really want to use the same (and if there are huge differences), then I'd vote for adding a separate alias file like `git-oh-my-zsh` to make it clear that these are the same aliases.

Out of curiosity: Why do you think that it's important that both use the same set of aliases?

**Comment by bittner** [comment]
A bit of background on my motivation:

I've only recently started working with short aliases. Before that I worked on automating the setup of development machines in a SWD agency roughly for the last year. Most of the developers were using Zsh when I joined, and new hires were forced to jump into it too, because it was a controlled, centralized setup without root access. I helped to make this more flexible, allowing a unified setup for all developers that gives freedom for - automated - custom setups, while still retaining centralized control for maintenance (but maintained from everyone in a DevOps fashion from then on). Hence, someone preferring a different distro, or Bash instead of Zsh, could have that easily too, by adding some custom adaptions to the configuration repositories.

Personally, I prefer having a setup that is as close to a distro default as possible, and ideally mainstream (which translates to Ubuntu and Bash, sorry folks!). Also, the wonky magic of the Oh My Zsh prompt made me feel uncomfortable. - _But that may all be just a question of taste and getting used to it!_

## TL;DR

> Before you embark on this, could you outline the main differences?

I've not really made an analysis yet. When you click on the two links above you'll see that - from a first impression - On My Zsh has more. Some are also different (e.g. `gs` vs. `gst`, which makes sense when you use Ghostscript :smirk:).

> Why do you think that it's important that both use the same set of aliases?

Three reasons:

1. It's **easier to switch** from Zsh to Bash (and vice-versa). No excuses anymore.
1. When developers do **pair programming** they can use the same set of short Git commands, and they can learn from each other, even though one may be using Zsh and the other Bash.
1. I promote employing best practices in SWD. And best practices are things that are popular. As Oh My Zsh seems to be tremendously popular already it makes sense, in my eyes, to **build on their popularity** aligning with them (which somewhat brings us back to item no. 1).

Bottom line: I want (to provide) more freedom.

## What we could do

> adding a separate alias file

I'll give that a shot. You're right, we shouldn't punish the muscle memory of people using Bash-it today. Therefore, I'll try to find out

- which are the aliases common to both Bash-it and Oh My Zsh
- which aliases we would have to add (i.e. they currently exist only in Oh My Zsh)
- which aliases are "proprietary" to Bash-it
- where there are currently conflicts

and put that in a PR, if it makes sense.

How does that sound?

## See also

You may be interested to know that I'm also talking to the Oh My Zsh project:

- https://github.com/robbyrussell/oh-my-zsh/issues/6793 - _"Add `gav` as git-alias for `git add -v`"_

**Comment by nwinkler** [comment]
Wow - thanks for the detailed reply! Yes, that makes a lot of sense. Please go ahead with this...

Are you planning to duplicate the common aliases in each of the files (`git` and `git-oh-my-zsh`), or putting them in a common file? The latter case would be cleaner, but I'm not sure how they would be loaded in this case, since Bash-it currently does not have an out-of-the-box way of importing an alias file from another alias file...

**Comment by bittner** [comment]
I would like to try coming up with just one configuration in the long run (i.e. avoid a "run Bash-it in Oh-My-Zsh mode" feature), so it's the latter case, the common file for now.

Maybe I can just have a single file with the Oh My Zsh configuration on top and the current Bash-it differences at the bottom, thus overriding aliases where there are conflicts. But I have to figure that out. The Oh My Zsh configuration seems to have constructs proprietary to Zsh that make taking things over repeatedly not a totally trivial task.

Let's talk through the details once I have a better idea about what the differences really are.

**Comment by bittner** [comment]
> Bash-it currently does not have an out-of-the-box way of importing an alias file

Sourcing a file seems to be possible, though it looks and feels a bit ugly. I've tried putting Oh My Zsh's Git alias plugin into a `./aliases/compat/` folder and source it, like this:

```bash
cite 'about-alias'
about-alias 'common git abbreviations'

BASHIT_HOME="$(cd -P "$(dirname "${BASH_SOURCE[0]}")/.." && pwd)"
source "${BASHIT_HOME}/aliases/compat/git.plugin.zsh"

# Aliases
alias gcl='git clone'
...
```

Is there a better way?

**Comment by nwinkler** [comment]
The `BASH_IT` variable should be set to the install location of Bash-it. You shouldn't have to do the `BASHIT_HOME` dance that you did there, just use `${BASH_IT}` instead.

**Comment by bittner** [comment]
Done.

How does the autocompletion work? Can we get this working easily with the additional aliases coming from Oh My Zsh?

**Comment by bittner** [comment]
Hi @nwinkler, how was your summer season?

Can you shed light into my 2 questions from above (autocompletion, aliases-help) to help us figure out if PR #1201 is complete as it is, or if we can make it more useful?

**Comment by nwinkler** [comment]
Sorry for the late response, @bittner... Yes, summer vacation was very nice, thanks for asking.

To be honest, I don't know off the top of my head whether your code would be causing any issues with these two cases. I had a quick look at the code that generates the help for the aliases (in `lib/helpers.bash`), and it looks like the current code is simply using `cat` to pipe the alias file through a couple of filters. Based on this, I don't think that the new aliases (sourced) will be showing up, since the `cat` does not source the referenced file. This will probably require some changes...

The autocomplete will probably work, since it uses the aliases that are defined in the current shell.

Best strategy for verifying this:
* Give it a manual try to see whether the newly sourced aliases show up in autocomplete and `bash-it help aliases git`.
* Please also add/extend the tests in `test/lib/helpers.bats` to verify that the sourced aliases are supported in all cases.

Sorry if this does not help much...

**Comment by ctrngk** [comment]
when I add 
plugins=(python)
in  .zshrc of oh-my-zsh
it also includes some useful functions like pyclean, which is to clean cache python files in the directory.

This is only an example of what oh-my-zsh has provided. I only use plugin feature in oh-my-zsh (which is alias in bash-it). Now I am leaving to oh-my-bash project which seems like a exact copy of oh-my-zsh

**Comment by NoahGorny** [comment]
hey @bittner 
do you still want to do it?

**Comment by bittner** [comment]
Actually, this would be a super-important addition. I still have this fix on my local machine.

Unless there is a _very_ good reason not to do so, the short aliases of Bash-it and Oh My Zsh should be align as much as possible, I feel. Motivations are [mentioned above](https://github.com/Bash-it/bash-it/issues/1191#issuecomment-392462427).

**Comment by NoahGorny** [comment]
> Actually, this would be a super-important addition. I still have this fix on my local machine.
> 
> Unless there is a _very_ good reason not to do so, the short aliases of Bash-it and Oh My Zsh should be align as much as possible, I feel. Motivations are [mentioned above](https://github.com/Bash-it/bash-it/issues/1191#issuecomment-392462427).

I see, I think that in that case, we should see what is the difference, and either create a new `zsh-git` completion file, or just change ours. I think that adding a new file would be ideal, and willing to do so provided someone would research the difference between us and oh-my-zsh.
We would also need to write it done somewhere in the documentation in order to add new git aliases in the correct place.

Thanks for opening it up again @bittner, sorry for closing it in the first place!

**Comment by cornfeedhobo** [comment]
@NoahGorny The zsh aliases can be seen [here](https://github.com/ohmyzsh/ohmyzsh/blob/master/plugins/git/git.plugin.zsh)

**Comment by gaelicWizard** [comment]
> I think that adding a new file would be ideal

I suggest that having a single file would be ideal. I agree with @bittner that having similarity and building on popular usage are valuable in and of themselves. Its weird to do "_Bash It_ in _Oh My Zsh_ compatibility mode"... I'd suggest even disabling some old aliases, like `gs`, for new users. $0.02

**Comment by NoahGorny** [comment]
I just think that bash-it has his own flavours of aliases, and I wouldn't like to throw them all away. Its better to leave the option for people to choose for themselves between the bash-it ones and ohmyzsh ones

**Comment by davidpfarrell** [comment]
> I'd suggest even disabling some old aliases, like `gs`

I've already used my `gs` alias over 100 times today, so I vote to keep it :)

> I just think that bash-it has his own flavours of aliases

I completely agree.

I also don't see zsh as being any particular authority on aliases in general.

side note: I do wish there was a central website for shell alias discussion / best practices - then we could introduce plugins that align with them.

As it is, I think adding a new separate plugin that ports the zsh git aliases as much as possible to bash, would be useful to many people.  It would essentially exist to be a proper bashified port of the zsh aliases and would not stray from that mandate.

That leaves the current git aliases to grow and evolve on their own, as seen fit by the bash-it community.

**Comment by gaelicWizard** [comment]
Should this issue be closed now that #1831 is merged?

**Comment by NoahGorny** [comment]
Unfortunately, its not merged yet- I still need to get around and address your comments and other things as well!

**Comment by gaelicWizard** [comment]
Oy! All this merging and I jumped the gun!!

**Comment by bittner** [comment]
> side note: I do wish there was a central website for shell alias discussion / best practices - then we could introduce plugins that align with them.

Why not start such a project? – I did some research on cool, available domain names. 😄

(Sorted by current price 💲; rounded up ⤴; some cheap US domain dealer 🇺‍🇸.)

## Theme: Health / Wellness / Beauty / Thinner / Less

| Domain | Buy |Renew | Note |
| --------- |-------|------ |---- |
| git.makeup | $2 | $12 |  |
| git.fitness | $6 | $29 |  |
| git.doctor | $8 | $83 |  |
| git.diamonds | $40 | $40 |  |
| git.yoga | $42 | $25 | <kbd>premium</kbd> |
| git.jewelry | $47 | $47 |  |
| git.diet |$94 | $104 |  |

## Theme: Fix broken things / Create stuff

| Domain | Buy |Renew | Note |
| --------- |-------|------ |---- |
| git.repair | $6 | $26 |  |
| git.enterprises | $6 | $28 |  |
| git.parts | $10 | $29 |  |
| git.supply | $16 | $19 |  |
| git.supplies | $18 | $18 |  |
| git.construction | $28 | $28 |  |
| git.plumbing | $52 | $52 |  |
| git.how | $351 | $351 | <kbd>premium</kbd> |
| git.diy | $517 | $1,035 | <kbd>premium</kbd> |

## Theme: Bargain / Great offers / Fine dining

| Domain | Buy |Renew | Note |
| --------- |-------|------ |---- |
| git.poker | $8 | $48 |  |
| git.promo | $11 | $18 |  |
| git.forsale | $11 | $26 |  |
| git.restaurant | $11 | $46 |  |
| git.gratis | $17 | $17 |  |
| git.discount | $24 | $24 |  |
| git.bingo | $43 | $43 |  |

## Theme: Nerds Zone  <kbd>_**NSFW**_</kbd>

| Domain | Buy |Renew | Note |
| --------- |-------|------ |---- |
| git.porn | $88 | $88 | potentially <kbd>NSFW</kbd> |
| git.love | $2'070 | $4'140 | potentially <kbd>NSFW</kbd> |

----

Does anyone feel like starting a sweet website and we try to kick off some kind of collaboration between shell projects? 🚀

**Comment by seefood** [comment]
@bittner I fail to see how this last message is relevant or on topic.

**Comment by bittner** [comment]
> [@bittner](https://github.com/bittner) I fail to see how this last message is relevant or on topic.

It's a direct response to https://github.com/Bash-it/bash-it/issues/1191#issuecomment-898958652. @davidpfarrell's note was that solving the Git aliases alignment problem would be easier with a central place to consent about Git aliases. Make sense?

**Comment by seefood** [comment]
not really. you can open a repo, there are even tools for cross-shell alias definitions, no need to spam here with the prices of git.porn and other domains. this not the place.

have a look at https://github.com/nomyfan/alias-rs
and I have seen at least one other such project

**Comment by bittner** [comment]
Just for the record, #1831 is still unmerged. 🚧 🏗

**Comment by seefood** [comment]
ok, fair enough :)
there's an unsettled discussion there.
